includes/handler.inc

```diff
@@ -447,71 +454,107 @@ function islandora_oai_query_solr($start = 0, $field = 'PID', $query = NULL, $se
+        $xacml_filters = array(
+          '!bound(?role) && !bound(?user)',
+          'bound(?role) && (' . implode(' || ', $role_filters) . ')',
+          "bound(?user) && ?user = '$user_name'",
+        );
+        $parenthesis_wrap = function($string) {
+          return "($string)";
+        };
+        $xacml_filter = implode(' || ', array_map($parenthesis_wrap, $xacml_filters));
+        $query = <<<EOQ
+SELECT ?child
+FROM <#ri>
+WHERE {
+  ?child <fedora-model:state> <fedora-model:Active> .
+  {
+    ?child <fedora-model:hasModel> <info:fedora/islandora:collectionCModel>
+  } UNION {
+    ?child <fedora-model:hasModel> [<fedora-model:hasModel> <info:fedora/islandora:collectionCModel>]
+  } .
+  {
+    ?child <fedora-rels-ext:isMemberOfCollection>+ <info:fedora/$value>
+  } UNION {
+    ?child <fedora-rels-ext:isMemberOf>+ <info:fedora/$value>
+  }
+  OPTIONAL {
+    ?child <http://islandora.ca/ontology/relsext#isViewableByRole> ?role .
+  }
+  OPTIONAL {
+    ?child <http://islandora.ca/ontology/relsext#isViewableByUser> ?user .
+  }
+  FILTER($xacml_filter)
+}
+EOQ;
+        $query_results = $connection->repository->ri->sparqlQuery($query);
+      }
+      else {
+        $itql_query = '
+        select $child from <#ri>
+        where
         (
           (
-          $child <fedora-model:hasModel> $model
-          and $model <mulgara:is> <info:fedora/islandora:collectionCModel>
-          and $parent_model <mulgara:is> <info:fedora/fedora-system:ContentModel-3.0>
+            (
+            $child <fedora-model:hasModel> $model
+            and $model <mulgara:is> <info:fedora/islandora:collectionCModel>
+            and $model <fedora-model:hasModel> <info:fedora/fedora-system:ContentModel-3.0>
+            )
+          or
+            (
+            $child <fedora-model:hasModel> $model
+            and $model <fedora-model:hasModel> $parent_model
+            and $parent_model <mulgara:is> <info:fedora/islandora:collectionCModel>
+            and $child <fedora-model:state> <info:fedora/fedora-system:def/model#Active>
+            )
+          minus $child <http://islandora.ca/ontology/relsext#isViewableByRole> $role
+          minus $child <http://islandora.ca/ontology/relsext#isViewableByUser> $user
           )
         or
           (
-          $child <fedora-model:hasModel> $model
-          and $model <fedora-model:hasModel> $parent_model
-          and $parent_model <mulgara:is> <info:fedora/islandora:collectionCModel>
-          and $child <fedora-model:state> <info:fedora/fedora-system:def/model#Active>
-          )
-        minus $child <http://islandora.ca/ontology/relsext#isViewableByRole> $role
-        minus $child <http://islandora.ca/ontology/relsext#isViewableByUser> $user
+            (
+            $child <fedora-model:hasModel> $model
+            and $model <mulgara:is> <info:fedora/islandora:collectionCModel>
+            and $parent_model <mulgara:is> <info:fedora/fedora-system:ContentModel-3.0>
```

Comment:
A note to myself as I notice it: Another oddly bound/unbound `$parent_model` dealio, making no real assertion.